docs/Templates-Testing-Tooling.md

```diff
@@ -1 +1,287 @@
+
+## Naming of snapshots
+
+All snapshots are stored in a single folder that is by default named `Snapshots` and is located in a folder with test code (in case of API usage) or current folder (in case of CLI) usage. Location of the snapshots directory can be optionaly specified in API via [SnapshotsDirectory](#SnapshotsDirectory) parameter or in CLI via [-d|--snapshots-directory](#--snapshots-directory) option.
+
+All the files consisting a single snapshot of single scenario are stored in a single folder that is named with following convention: 
+
+```
+<caller_method>.<template_name>.<user_passe_scenario_name>.<encoded_input_args>.<scenario_differentiator>.<received|verified>
+```
+
+ - `<caller_method>` - not used in CLI. Added by default in API. Can be opted out via [`DoNotPrependCallerMethodNameToScenarioName`](#DoNotPrependCallerMethodNameToScenarioName)
+
+ - `<template_name>` - used by default. Can be opted out in API via [`DoNotPrependTemplateNameToScenarioName`](#DoNotPrependTemplateNameToScenarioName)
+
+  - `<user_passe_scenario_name>` - not used by default. Can be opted in in API via specifying [`ScenarioName`](#ScenarioName)
+
+   - `<encoded_input_args>` - used by default. Can be opted out in API via [`DoNotAppendTemplateArgsToScenarioName`](#DoNotAppendTemplateArgsToScenarioName)
+
+  - `<scenario_differentiator>` - not used by default. Can be opted in in CLI via [`--unique-for`](#--unique-for) and in API via [`UniqueFor`](#UniqueFor)
+
+  - `<received|verified>` - suffix, depending on whether the snapshot is a basline for test scenario (then the suffix is '.verified') or created by engine as unexpected/regressing the check (then the suffix is '.received').
+
+
+## Layout of snapshot files
+
+All files within snapshot has identical filenames and relative folder structure as files in the generated template output.
+
+Snapshot folder has the following layout:
+
+```
+<Snapshot_folder_name.received|verified>
+  |--<template_name>
+  |   |--<file_1 from generated template>
+  |   |--<file_2 from generated template>
+  |   |-- ... // layout is respecting folder structure of generated template
+  |
+  |--<std-streams> // present if opted in into verifying command output
+      |--stderr.txt
+      |--stdout.txt
+```
+
+# API
+
+## Setup
+
+1. Add the package reference to `Microsoft.TemplateEngine.Authoring.TemplateVerifier`:
+  ```dotnetcli
+  dotnet add package Microsoft.TemplateEngine.Authoring.TemplateVerifier
+  ```
+2. Start using the `TemplateVerifier` in your tests:
+  ```cs
+using Microsoft.TemplateEngine.Authoring.TemplateVerifier;
+
+namespace MyTests
+
+[TestClass]
+public class MyTestClass
+{
+    [TestMethod]
+    public async Task MyTemplate_InstantiationTest()
+    {
+        TemplateVerifierOptions options = new(templateName: templateShortName)
+        {
+            TemplateSpecificArgs = new[] { "--arg1", "true" },
+        };
+        
+        VerificationEngine engine = new(_logger);
```

Comment:
it would be good to have an overload from XUnit logging too

Reply:
Currently the API is intentionaly agnostic on the testing frameworks (under the hood it uses `Verify.Xunit`, but just because the Verify functionality is currently surfaced only through 'protocol specific packages', however XUnit is neither enforced nor surfaced anywhere).

E.g. this specific code sample is in `MSTest`